wallet/src/Pos/Util/Mnemonic.hs

```diff
@@ -0,0 +1,664 @@
+-- | Module providing restoring from backup phrase functionality
+
+module Pos.Util.Mnemonic
+       (
+       -- * Types
+         Mnemonic
+       , Entropy
+
+       -- * Creating @Mnemonic@ (resp. @Entropy@)
+       , mkEntropy
+       , mkMnemonic
+
+       -- * Converting from and to @Mnemonic@ (resp. @Entropy@)
+       , mnemonicToEntropy
+       , mnemonicToSeed
+       , mnemonicToAESKey
+       , entropyToMnemonic
+       , entropyToByteString
+       ) where
+
+import           Universum
+
+import           Crypto.Hash (Blake2b_256, Digest, SHA256, hash)
+import           Data.Aeson (FromJSON (..), ToJSON (..))
+import           Data.Aeson.Types (Parser)
+import           Data.Bits (Bits, shiftL, shiftR, (.&.))
+import           Data.ByteArray (convert)
+import           Data.ByteString (ByteString)
+import           Data.Char (isAscii)
+import           Data.Default (Default (def))
+import           Data.List (elemIndex, (!!))
+import           Data.Swagger (NamedSchema (..), ToSchema (..))
+import           Data.Text.Buildable (Buildable (build))
+import           Test.QuickCheck (Arbitrary (arbitrary))
+import           Test.QuickCheck.Gen (oneof, vectorOf)
+
+import           Pos.Binary (serialize')
+import           Pos.Util.LogSafe (SecureLog)
+
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Char8 as B8
+import qualified Data.Text as Text
+
+
+-- | A backup-phrase in the form of a non-empty of Mnemonic words
+-- Constructor isn't exposed.
+data Mnemonic = Mnemonic
```

Comment:
Note that this might be useful when we add redeem functions to V1 - to distinguish between mnemonics of different length

Reply:
I thought of this. And ideally, I'd even like to have different types for them, and define the conversion functions as type-class. Otherwise, there's nothing really preventing us to use a `Mnemonic12` in a context where we want a `Mnemonic15`. Having the number of words at the type level with a `Nat` could do the trick. I'll spend some time today to see if I can get something to work and if the complexity it brings isn't too prevalent.